### PR TITLE
[s4u] add mutex

### PR DESCRIPTION
**Pull request by adfaure**
An example can be found at  `examples/s4u/actor/`.

### CODE REVIEW

File: include/simgrid/s4u/mutex.hpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ * Rendez-vous point for network communications, similar to URLs on which you could post and retrieve data.
+ * They are not network locations: you can post and retrieve on a given mailbox from anywhere on the network.
+ * You can access any mailbox without any latency. The network delay are only related to the location of the
+ * sender and receiver.
+ */
+XBT_PUBLIC_CLASS Mutex {
+
+public:
+  Mutex();
+  ~Mutex() {};
```
Comment (randomstuff) [inline include/simgrid/s4u/mutex.hpp]: We should destroy `_mutex` here.

File: include/simgrid/s4u/mutex.hpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+namespace s4u {
+
+/** @brief Mailboxes
+ *
+ * Rendez-vous point for network communications, similar to URLs on which you could post and retrieve data.
+ * They are not network locations: you can post and retrieve on a given mailbox from anywhere on the network.
+ * You can access any mailbox without any latency. The network delay are only related to the location of the
+ * sender and receiver.
+ */
+XBT_PUBLIC_CLASS Mutex {
```
Comment (randomstuff) [inline include/simgrid/s4u/mutex.hpp]: If `Mutex` implements unique onwership of the underlying mutex it should not be copyable but moveable.

Comment (randomstuff) [inline include/simgrid/s4u/mutex.hpp]: Maybe we should try to follow the [Mutex](http://en.cppreference.com/w/cpp/concept/Mutex) concept: non copyable, non moveable.

Comment (randomstuff) [inline include/simgrid/s4u/mutex.hpp]: In fact, we only need the `Lockable` concept to make it work with `lock_guard` and other.

File: include/simgrid/s4u/mutex.hpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+#define SIMGRID_S4U_MUTEX_HPP
+
+#include <xbt/base.h>
+#include "simgrid/simix.h" 
+
+
+namespace simgrid {
+namespace s4u {
+
+/** @brief Mailboxes
```
Comment (randomstuff) [inline include/simgrid/s4u/mutex.hpp]: This one is wrong :)

File: src/s4u/s4u_mutex.cpp [hunk 4]
```diff
@@ -14,17 +14,18 @@
 using namespace simgrid;
 
 s4u::Mutex::Mutex() {
-    _mutex = simcall_mutex_init();  
+    smx_mutex_t smx_mutex = simcall_mutex_init();  
+    _mutex = std::shared_ptr<s_smx_mutex>(smx_mutex, [](smx_mutex_t mutex){ SIMIX_mutex_destroy(mutex); });
```
Comment (randomstuff) [inline src/s4u/s4u_mutex.cpp]: In fact, we can do this:

    _mutex = std::shared_ptr<s_smx_mutex>(smx_mutex, SIMIX_mutex_destroy);

which is probably simplet :)

File: examples/s4u/actor/s4u_actor.cpp [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  }
+};
+
+class MainActor {
+public:
+  void operator()() {
+    int *res = new int;
+    simgrid::s4u::Mutex *mtx = new simgrid::s4u::Mutex();
+    *res = 0;
+    simgrid::s4u::Actor* workers[NB_ACTOR*2];
```
Comment (randomstuff) [inline examples/s4u/actor/s4u_actor.cpp]: I get this warning:

    simgrid/examples/s4u/actor/s4u_actor.cpp:68:26: error: variable ‘workers’ set but not used [-Werror=unused-but-set-variable]

File: examples/s4u/actor/s4u_actor.cpp [hunk 6]
```diff
@@ -0,0 +1,87 @@
+/* Copyright (c) 2006-2015. The SimGrid Team. All rights reserved.          */
```
Comment (randomstuff) [inline examples/s4u/actor/s4u_actor.cpp]: The name of this test should probably be changed to include "mutex" inside it if the tests is supposed to be about mutexes.

File: examples/s4u/actor/s4u_actor.cpp (hunk 5 above)
Comment (adfaure) [inline examples/s4u/actor/s4u_actor.cpp]: Yes, initially I wanted to join those actors.

**Comment by randomstuff** [comment]
Merged (with a small fix).

**Pull request closed without merging**